### DISCUSSION TITLE
Rack 3 fails where rack 2.2 worked

### DISCUSSION THREAD

**Discussion by DannyBen** [rack/rack#1977] [Q&A]
Hi all,

I was trying to find out what I am doing wrong here.
It seems that something changed in Rack 3 that makes the below code fail.

The file is ready to be executed as is with `ruby server.rb`.
I intentionally avoid using `rackup` (since this code below is a simplified version of [this server](https://github.com/DannyBen/httpme/blob/master/lib/httpme/server.rb)).

```ruby
# server.rb
require 'bundler/inline'

gemfile do
  source "https://rubygems.org"

  # gem "rack", "= 3.0.0"  # breaks
  gem "rack", "= 2.2.4"    # works
  gem "puma", "= 6.0.0"
end

require 'rack'
require 'rack/handler/puma'

class Server
  def run
    Rack::Handler::Puma.run app, Port: '3000', Host: '0.0.0.0'
  end

  def app
    Proc.new do |env|
      [200, {}, ["Hello World"]]
    end
  end
end

Server.new.run
```

With rack 2.2 it works, with rack 3 it outputs this error:
**cannot load such file -- rack/handler (LoadError)**

```
➜ ruby server.rb
/home/vagrant/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/rack/handler/puma.rb:3:in `require': cannot load such file -- rack/handler (LoadError)
        from /home/vagrant/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/puma-6.0.0/lib/rack/handler/puma.rb:3:in `<top (required)>'
        from server.rb:12:in `require'
        from server.rb:12:in `<main>'
```

Can anyone point me in the right direction?

**Comment by ioquatix** [comment]
Yes you need to take a look at the new gem, called rackup.

**Reply by DannyBen** [reply]
But...

1. I do not want to use the `rackup` CLI.
2. Installing `rackup` gem conflicts with rack < 3 - and rack 2.x is still required by many other apps.
3. Puma [still places its file in `rack/handler` and not `rackup/handler`](https://github.com/puma/puma/tree/master/lib/rack/handler)

**Reply by ioquatix** [reply]
In Rack 2.x, the `rackup` CLI was included regardless of whether you wanted it, and now it's optional (in a separate gem). The server and handler code is not strictly part of "Rack the interface" that's why it's now independent. The `rackup` gem has compatibility files in place for `rack/handler` so it continues to work.

I understand this isn't exactly the answer you want, so feel free to help shape the future of Rack and Rackup by proposing a better way forward. If it makes sense, I don't see why we wouldn't consider it.

**Reply by DannyBen** [reply]
Well, first of all, thanks for the discussion.

Since I am not involved in rack development, and the decisions and discussions that led to this situation, I am not in any position to offer something better. That said, I think there are problems with the current situation.

Even when including `rackup` in the above example, I am getting a similar error, now from the rackup gem:

```
➜ ruby server.rb
/home/vagrant/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/rackup-0.1.0/lib/ra
ckup/handler.rb:3:in `require': cannot load such file -- rack/handler (LoadError)
        from /home/vagrant/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/rackup-0.1.0/lib/rackup/handler.rb:3:in `<top (required)>'
        from /home/vagrant/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/rackup-0.1.0/lib/rackup.rb:1:in `require_relative'
        from /home/vagrant/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/rackup-0.1.0/lib/rackup.rb:1:in `<top (required)>'
```

This even happens with just bundling and doing nothing else:

```ruby
# server.rb
require 'bundler/inline'

gemfile do
  source "https://rubygems.org"

  gem "rack", "= 3.0.0"
  gem "puma", "= 6.0.0"
  gem "rackup"
end

require 'rackup/handler/puma'


```

and it is only passing the error after installing webrick.
After that, the above file still fails with

`cannot load such file -- rackup/handler/puma (LoadError)`

replacing it to `rack/handler...` shows a deprecation error...


---

It feels like I am chasing a ghost.
